### PR TITLE
Added the functions to Cas9 Stats

### PR DESCRIPTION
**Pull request by Salahub**
These functions generate plots and determine frequencies for different characteristics of the mutants, such as indices and amino acid changes

### CODE REVIEW

File: models/tridimensional/data_kleinstiver/cas9_mutants_stats.py [hunk 1]
```diff
@@ -137,3 +137,106 @@ def main():
...
+    plt.xticks(range(len(ch)), [aa for aa in ch.keys()], rotation=45, ha='right')
+    plt.title("Mutations in Amino Acids of Successful " + pam + " Cas9 Mutants")
+    plt.xlabel("Amino Acid Category Change")
+    plt.ylabel("Mutation Count")
+    plt.yscale('log')
+    plt.legend()
+    plt.show()
+    
+bar_chn_counts('NGA')
+bar_chn_counts('NGC')
```
Comment (mattsmart) [inline models/tridimensional/data_kleinstiver/cas9_mutants_stats.py]: these should prob be put in the main call

File: models/tridimensional/data_kleinstiver/cas9_mutants_stats.py [hunk 2]
```diff
@@ -137,3 +137,106 @@ def main():
...
+
+def ss_freq(pam):
+    '''
+    Given a pam site returns the secondary structures present in the mutations
+    '''
+    
+    mut_lst = []
+    for mutant in mutants_kleinstiver:
+        if mutant['pam'] == pam:
+            mut_lst.append(mutant)
```
Comment (mattsmart) [inline models/tridimensional/data_kleinstiver/cas9_mutants_stats.py]: `mut_lst = [mutant for mutant in mutants_kleinstiver if (mutant['pam'] == pam)]`

this abbreviates these 4 lines and is faster

File: models/tridimensional/data_kleinstiver/cas9_mutants_stats.py [hunk 3]
```diff
@@ -137,3 +137,106 @@ def main():
...
+                id_dict[idx] = id_dict[idx] + 1
+            if not idx in id_dict.keys():
+                id_dict[idx] = 1
+    return id_dict
+
+def ss_freq(pam):
+    '''
+    Given a pam site returns the secondary structures present in the mutations
+    '''
+
```
Comment (mattsmart) [inline models/tridimensional/data_kleinstiver/cas9_mutants_stats.py]: no need for extra line

File: models/tridimensional/data_kleinstiver/cas9_mutants_stats.py [hunk 4]
```diff
@@ -137,3 +137,106 @@ def main():
...
+def idx_freq(pam):
+    '''
+    Given a pam site returns the indices of mutation and frequency of mutation
+    occurring at said indices
+    '''
+    
+    mut_lst = []
+    for mutant in mutants_kleinstiver:
+        if mutant['pam'] == pam:
+            mut_lst.append(mutant)
```
Comment (mattsmart) [inline models/tridimensional/data_kleinstiver/cas9_mutants_stats.py]: `mut_lst = [mutant for mutant in mutants_kleinstiver if (mutant['pam'] == pam)]`

this abbreviates these 4 lines and is faster

File: models/tridimensional/data_kleinstiver/cas9_mutants_stats.py [hunk 5]
```diff
@@ -137,3 +137,106 @@ def main():
...
+def aachng_freq(pam):
+    '''
+    Given a pam site returns the indices and change of amino acids in each mutation in the form
+    aa_init to aa_fin at aa_idx
+    '''
+    
+    mut_lst = []
+    for mutant in mutants_kleinstiver:
+        if mutant['pam'] == pam:
+            mut_lst.append(mutant)
```
Comment (mattsmart) [inline models/tridimensional/data_kleinstiver/cas9_mutants_stats.py]: `mut_lst = [mutant for mutant in mutants_kleinstiver if (mutant['pam'] == pam)]`

this abbreviates these 4 lines and is faster

File: models/tridimensional/data_kleinstiver/cas9_mutants_stats.py [hunk 6]
```diff
@@ -137,3 +137,106 @@ def main():
...
+
+def aafin_freq(pam):
+    '''
+    Given a pam site returns the final amino acid of mutation
+    '''
+    
+    mut_lst = []
+    for mutant in mutants_kleinstiver:
+        if mutant['pam'] == pam:
+            mut_lst.append(mutant)
```
Comment (mattsmart) [inline models/tridimensional/data_kleinstiver/cas9_mutants_stats.py]: `mut_lst = [mutant for mutant in mutants_kleinstiver if (mutant['pam'] == pam)]`

this abbreviates these 4 lines and is faster

File: models/tridimensional/data_kleinstiver/cas9_mutants_stats.py (hunk 1 above)
Comment (mattsmart) [inline models/tridimensional/data_kleinstiver/cas9_mutants_stats.py]: also make sure there's a newline at EOF

File: models/tridimensional/data_kleinstiver/cas9_mutants_stats.py [hunk 7]
```diff
@@ -137,3 +137,106 @@ def main():
...
+                aafin_dict[aafin] = 1
+    return aafin_dict
+
+
+def aachng_freq(pam):
+    '''
+    Given a pam site returns the indices and change of amino acids in each mutation in the form
+    aa_init to aa_fin at aa_idx
+    '''
+
```
Comment (mattsmart) [inline models/tridimensional/data_kleinstiver/cas9_mutants_stats.py]: remove extra line

**Comment by mattsmart** [comment]
comments: minor style things and some python idiosyncrasies

also would be good to make sure the main is the last function and main is the last thing in the script (i.e. move all your functions above the main function)

i made one comment 4 times on the same block of code in diff places -- that block may be something thats worth creating its own function for

File: models/tridimensional/data_kleinstiver/cas9_mutants_stats.py [hunk 8]
```diff
@@ -137,3 +137,106 @@ def main():
...
+    mut_lst = []
+    for mutant in mutants_kleinstiver:
+        if mutant['pam'] == pam:
+            mut_lst.append(mutant)
+    aachng_dict = {}
+    for mutant in mut_lst:
+        for mutation in mutant['mutations']:
+            aachng = mutation['aa_group_init'] + ' to ' + mutation['aa_group_mut'] + ' at ' + str(mutation['aa_idx'])
+            if aachng in aachng_dict.keys():
+                aachng_dict[aachng] = aachng_dict[aachng] + 1
```
Comment (mattsmart) [inline models/tridimensional/data_kleinstiver/cas9_mutants_stats.py]: `aachng_dict[aachng] += 1`  is simpler

File: models/tridimensional/data_kleinstiver/cas9_mutants_stats.py [hunk 9]
```diff
@@ -137,3 +137,106 @@ def main():
...
+            if not aachng in aachng_dict.keys():
```
Comment (mattsmart) [inline models/tridimensional/data_kleinstiver/cas9_mutants_stats.py]: else:

File: models/tridimensional/data_kleinstiver/cas9_mutants_stats.py [hunk 10]
```diff
@@ -137,3 +137,106 @@ def main():
...
+def bar_chn_counts(pam):
+    """
+    Plots a single figure containing two bar charts showing the frequencies of mutations 
+    """
+    # Get Dictionary of counts
+    ch = aachng_freq(pam)
+    # Bar plot of secondary structure regions containing mutants in each mutant Cas9
+    bar_width = 0.45
+    plt.bar(range(len(ch)), [ch[aa] for aa in ch.keys()], width=bar_width, align='center',
+            color='#71cce6', label='NGA')
```
Comment (mattsmart) [inline models/tridimensional/data_kleinstiver/cas9_mutants_stats.py:194]: this had better be a rly nice colour

EDIT: just checked, not bad (unless tessa picked it then w.e)

File: models/tridimensional/data_kleinstiver/cas9_mutants_stats.py [hunk 11]
```diff
@@ -137,3 +137,106 @@ def main():
...
+            idx = mutation['aa_idx']
+            if idx in id_dict.keys():
+                id_dict[idx] = id_dict[idx] + 1
+            if not idx in id_dict.keys():
+                id_dict[idx] = 1
+    return id_dict
+
+def ss_freq(pam):
+    '''
+    Given a pam site returns the secondary structures present in the mutations
```
Comment (mattsmart) [inline models/tridimensional/data_kleinstiver/cas9_mutants_stats.py]: return the frequencies of mutation at different types of secondary structure

File: models/tridimensional/data_kleinstiver/cas9_mutants_stats.py [hunk 12]
```diff
@@ -137,3 +137,106 @@ def main():
...
 if __name__ == '__main__':
     main()
+
+    
+def idx_freq(pam):
+    '''
+    Given a pam site returns the indices of mutation and frequency of mutation
+    occurring at said indices
+    '''
+
```
Comment (mattsmart) [inline models/tridimensional/data_kleinstiver/cas9_mutants_stats.py]: remove extra line

File: models/tridimensional/data_kleinstiver/cas9_mutants_stats.py [hunk 13]
```diff
@@ -137,3 +137,106 @@ def main():
...
+        for mutation in mutant['mutations']:
+            aafin = mutation['aa_group_mut']
+            if aafin in aafin_dict.keys():
+                aafin_dict[aafin] = aafin_dict[aafin] + 1
+            if not aafin in aafin_dict.keys():
+                aafin_dict[aafin] = 1
+    return aafin_dict
+
+
+def aachng_freq(pam):
```
Comment (mattsmart) [inline models/tridimensional/data_kleinstiver/cas9_mutants_stats.py]: aachng is not a very descriptive name, aa_change or aa_swap would be better

File: models/tridimensional/data_kleinstiver/cas9_mutants_stats.py [hunk 14]
```diff
@@ -137,3 +137,106 @@ def main():
...
+        for mutation in mutant['mutations']:
+            aachng = mutation['aa_group_init'] + ' to ' + mutation['aa_group_mut'] + ' at ' + str(mutation['aa_idx'])
+            if aachng in aachng_dict.keys():
+                aachng_dict[aachng] = aachng_dict[aachng] + 1
+            if not aachng in aachng_dict.keys():
+                aachng_dict[aachng] = 1
+    return aachng_dict
+
+
+def bar_chn_counts(pam):
```
Comment (mattsmart) [inline models/tridimensional/data_kleinstiver/cas9_mutants_stats.py]: `histogram_mutation_frequency_by_aa_???(pam)`

chn is not descriptive and different from chng lol

File: models/tridimensional/data_kleinstiver/cas9_mutants_stats.py [hunk 15]
```diff
@@ -137,3 +137,106 @@ def main():
...
+    aafin_dict = {}
+    for mutant in mut_lst:
+        for mutation in mutant['mutations']:
+            aafin = mutation['aa_group_mut']
+            if aafin in aafin_dict.keys():
+                aafin_dict[aafin] = aafin_dict[aafin] + 1
+            if not aafin in aafin_dict.keys():
+                aafin_dict[aafin] = 1
+    return aafin_dict
+
```
Comment (mattsmart) [inline models/tridimensional/data_kleinstiver/cas9_mutants_stats.py:166]: this block has identical structure to two*\* above, and very similar to one below -- might be good to functionalize

File: models/tridimensional/data_kleinstiver/cas9_mutants_stats.py [hunk 16]
```diff
@@ -137,3 +137,106 @@ def main():
...
+    loc_dict = {}
+    for mutant in mut_lst:
+        for mutation in mutant['mutations']:
+            loc = mutation['sec_structure']
+            if loc in loc_dict.keys():
+                loc_dict[loc] = loc_dict[loc] + 1
+            if not loc in loc_dict.keys():
+                loc_dict[loc] = 1
+    return loc_dict
+
```
Comment (mattsmart) [inline models/tridimensional/data_kleinstiver/cas9_mutants_stats.py:150]: this block has identical structure to several others -- might be good to functionalize

File: models/tridimensional/data_kleinstiver/cas9_mutants_stats.py [hunk 17]
```diff
@@ -137,3 +137,106 @@ def main():
...
+        if mutant['pam'] == pam:
+            mut_lst.append(mutant)
+    id_dict = {}
+    for mutant in mut_lst:
+        for mutation in mutant['mutations']:
+            idx = mutation['aa_idx']
+            if idx in id_dict.keys():
+                id_dict[idx] = id_dict[idx] + 1
+            if not idx in id_dict.keys():
+                id_dict[idx] = 1
```
Comment (mattsmart) [inline models/tridimensional/data_kleinstiver/cas9_mutants_stats.py:132]: this block has identical structure to several others -- might be good to functionalize

**Comment by mattsmart** [comment]
looks good overall

comments: minor style things and some python idiosyncrasies

also would be good to make sure the main is the last function and main is the last thing in the script (i.e. move all your functions above the main function)

i made a few comments 4 times on the same block of code in diff places -- those blocks may warrnat new functions being made and called

File: models/tridimensional/data_kleinstiver/cas9_mutants_stats.py [hunk 18]
```diff
@@ -137,3 +137,106 @@ def main():
 
 if __name__ == '__main__':
```
Comment (jrhawley) [inline models/tridimensional/data_kleinstiver/cas9_mutants_stats.py]: should push this to the bottom of the file and put any function declarations before it

File: models/tridimensional/data_kleinstiver/cas9_mutants_stats.py (hunk 10 above)
Comment (alexanian) [inline models/tridimensional/data_kleinstiver/cas9_mutants_stats.py:194]: I did pick it, it's "iGEM Teal" from last year

**Comment by alexanian** [comment]
@Salahub can you push another commit with some of the changes requested? most especially pushing the **main** down and the smaller style changes, I can do the functionalization after we merge this

File: models/tridimensional/data_kleinstiver/cas9_mutants_stats.py [hunk 19]
```diff
@@ -137,3 +137,106 @@ def main():
...
+    for mutant in mut_lst:
+        for mutation in mutant['mutations']:
+            idx = mutation['aa_idx']
+            if idx in id_dict.keys():
+                id_dict[idx] = id_dict[idx] + 1
+            if not idx in id_dict.keys():
+                id_dict[idx] = 1
+    return id_dict
+
+def ss_freq(pam):
```
Comment (alexanian) [inline models/tridimensional/data_kleinstiver/cas9_mutants_stats.py:135]: This duplicates the function of find_ss_counts above almost exactly The only difference I can see is that find_ss_counts has keys for every secondary structure element (i.e. includes ones that are not mutated), unless I am very confused?

**Comment by Salahub** [comment]
Okay, I just added a file with those changes, I think. Tessa was mentioning she would work on the functionalization, so that was not changed as of yet.

**Comment by alexanian** [comment]
Thanks for the changes! I'll merge 'em and functionalize

**Comment by mattsmart** [comment]
this kills the .gitignore , gonna recover it lol

**Pull request merged**